src/dbnode/storage/index/block.go

```diff
@@ -249,258 +234,14 @@ func (b *block) EndTime() time.Time {
-	})
-	took := time.Since(start)
-	b.metrics.backgroundCompactionTaskRunLatency.Record(took)
-
-	if log {
-		logger.Debug("done compaction task", zap.Duration("took", took))
-	}
-
-	if err != nil {
-		return err
-	}
-
-	// Rotate out the replaced frozen segments and add the compacted one.
-	b.Lock()
-	defer b.Unlock()
-
-	result := b.addCompactedSegmentFromSegments(b.backgroundSegments,
-		segments, compacted)
-	b.backgroundSegments = result
-
-	return nil
-}
-
-func (b *block) addCompactedSegmentFromSegments(
-	current []*readableSeg,
-	segmentsJustCompacted []segment.Segment,
-	compacted segment.Segment,
-) []*readableSeg {
-	result := make([]*readableSeg, 0, len(current))
-	for _, existing := range current {
-		keepCurr := true
-		for _, seg := range segmentsJustCompacted {
-			if existing.Segment() == seg {
-				// Do not keep this one, it was compacted just then.
-				keepCurr = false
-				break
-			}
-		}
-
-		if keepCurr {
-			result = append(result, existing)
-			continue
-		}
-
-		err := existing.Segment().Close()
-		if err != nil {
-			// Already compacted, not much we can do about not closing it.
-			instrument.EmitAndLogInvariantViolation(b.iopts, func(l *zap.Logger) {
-				l.Error("unable to close compacted block", zap.Error(err))
-			})
-		}
-	}
-
-	// Return all the ones we kept plus the new compacted segment
-	return append(result, newReadableSeg(compacted, b.opts))
-}
-
 func (b *block) WriteBatch(inserts *WriteBatch) (WriteBatchResult, error) {
 	b.Lock()
+	defer b.Unlock()
 	if b.state != blockStateOpen {
-		b.Unlock()
 		return b.writeBatchResult(inserts, b.writeBatchErrorInvalidState(b.state))
 	}
-	if b.compact.compactingForeground {
-		b.Unlock()
-		return b.writeBatchResult(inserts, errUnableToWriteBlockConcurrent)
-	}
-	// Lazily allocate the segment builder and compactors
-	err := b.compact.allocLazyBuilderAndCompactors(b.blockOpts, b.opts)
-	if err != nil {
-		b.Unlock()
-		return b.writeBatchResult(inserts, err)
-	}
-
-	b.compact.compactingForeground = true
-	builder := b.compact.segmentBuilder
-	b.Unlock()
-
-	defer func() {
-		b.Lock()
-		b.compact.compactingForeground = false
-		b.cleanupForegroundCompactWithLock()
-		b.Unlock()
-	}()
-
-	builder.Reset(0)
-	insertResultErr := builder.InsertBatch(m3ninxindex.Batch{
-		Docs:                inserts.PendingDocs(),
-		AllowPartialUpdates: true,
-	})
-	if len(builder.Docs()) == 0 {
-		// No inserts, no need to compact.
-		return b.writeBatchResult(inserts, insertResultErr)
-	}
-
-	// We inserted some documents, need to compact immediately into a
-	// foreground segment from the segment builder before we can serve reads
-	// from an FST segment.
-	err = b.foregroundCompactWithBuilder(builder)
-	if err != nil {
-		return b.writeBatchResult(inserts, err)
-	}
-
 	// Return result from the original insertion since compaction was successful.
-	return b.writeBatchResult(inserts, insertResultErr)
+	return b.writeBatchResult(inserts, b.mutableSegments.WriteBatch(inserts))
```

Comment:
Ultimately I think it's good to move to a Lock()/defer Unlock() pattern here, however I don't think we can afford to hold onto the write lock while the foreground compaction happens (notice that before we release the write lock with "compactingForeground = true" and then reqacquire and set "compactingForeground = false" before this change).

We probably need a way to move the mutableSegments into "compactingForeground" phase while holding the lock, release it, then perform the write batch and foreground compaction, and then close out as "compactingForeground = false"  at the end.

What are your thoughts on how best to do this?

Reply:
Discussed offline, won't hold onto lock in the index block. Rely on locks in the `mutableSegment` struct. Locking logic there remains the same as it were when it was inside of the index block code.